### PR TITLE
Mean and std deviation for gas in fuzzing tests

### LINKED ISSUE #1300: Add `avg` and `sd` to fuzzing gas usage

### LINKED ISSUE #1339: Unify gas and fuzzer stats printing
`fuzzer_runs = 4321, gas: 131`
use either `:` or `=`

### PR DESCRIPTION
**Pull request by war-in**
Closes #1300
Closes #1339

## Introduced changes

- `GasStatistics` are extended with mean and std deviation

## Checklist

- [x] Linked relevant issue
- [x] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [x] Added changes to `CHANGELOG.md`

### CODE REVIEW

**Comment by war-in** [comment]
That's interesting, but we do not have any tests for fuzzer and gas 🤔

File: crates/forge-runner/src/test_case_summary.rs [hunk 1]
```diff
@@ -3,13 +3,46 @@ use crate::expected_result::{ExpectedPanicValue, ExpectedTestResult};
...
     pub min: u128,
     pub max: u128,
+    pub mean: Option<u128>,
+    pub std_deviation: Option<u128>,
+}
+
+impl GasStatistics {
+    pub fn new(gas_usages: &[u128]) -> Self {
+        let max = gas_usages.iter().copied().reduce(u128::max).unwrap();
+        let min = gas_usages.iter().copied().reduce(u128::min).unwrap();
```
Comment (cptartur) [inline crates/forge-runner/src/test_case_summary.rs:20]: Nit, I'd take `gas_usages` by value, so it's clear at the caller site that we need to clone this values

File: crates/forge-runner/src/test_case_summary.rs [hunk 2]
```diff
@@ -3,13 +3,46 @@ use crate::expected_result::{ExpectedPanicValue, ExpectedTestResult};
...
+            let sum_squared_diff = gas_usages
+                .iter()
+                .map(|&x| {
+                    let abs_difference = if x > mean { x - mean } else { mean - x };
+                    abs_difference.pow(2)
+                })
+                .sum::<u128>();
+
+            stats.mean = Some(mean);
+            stats.std_deviation = Some((sum_squared_diff / n).sqrt());
```
Comment (cptartur) [inline crates/forge-runner/src/test_case_summary.rs:39]: Let's just create mean, and std_deviation values and then return a new `GasStatistic` instance. I'd avoid mutable variables where possible

File: crates/forge/tests/data/fuzzing/src/lib.cairo [hunk 3]
```diff
@@ -116,4 +116,19 @@ mod tests {
...
+        loop {
+            if (i == a) {
+                break;
+            }
+
+            i += 1;
+        };
+
+        assert(1 == 1, '');
+    }
```
Comment (cptartur) [inline crates/forge/tests/data/fuzzing/src/lib.cairo:133]: Nice idea to test it like that

File: docs/src/testing/gas-etimation.md [hunk 4]
```diff
@@ -7,9 +7,18 @@ and gas-efficient.
 
 ## Displaying Estimated Gas
 
+### Single test
```
Comment (cptartur) [inline docs/src/testing/gas-etimation.md:10]:
```suggestion
### Single Test
```

File: docs/src/testing/gas-etimation.md [hunk 5]
```diff
@@ -7,9 +7,18 @@ and gas-efficient.
...
+
 When the test passes with no errors, estimated gas is displayed this way:
 ```shell
-[PASS] tests::simple_test, gas: ~1
+[PASS] tests::simple_test, gas = ~1
+```
+
+### Fuzzing tests
```
Comment (cptartur) [inline docs/src/testing/gas-etimation.md:17]:
```suggestion
### Fuzzed Tests
```

File: crates/forge-runner/src/test_case_summary.rs [hunk 6]
```diff
@@ -3,13 +3,49 @@ use crate::expected_result::{ExpectedPanicValue, ExpectedTestResult};
...
+use num_integer::Roots;
 use std::option::Option;
 
-#[derive(Debug, PartialEq, Clone)]
+#[derive(Debug, PartialEq, Clone, Default)]
 pub struct GasStatistics {
     pub min: u128,
     pub max: u128,
+    pub maybe_mean: Option<u128>,
+    pub maybe_std_deviation: Option<u128>,
```
Comment (Arcticae) [inline crates/forge-runner/src/test_case_summary.rs:14]:
```suggestion
    pub mean: Option<u128>,
    pub std_deviation: Option<u128>,
```
Type already contains this information

File: crates/forge-runner/src/test_case_summary.rs [hunk 7]
```diff
@@ -3,13 +3,49 @@ use crate::expected_result::{ExpectedPanicValue, ExpectedTestResult};
...
+}
+
+impl GasStatistics {
+    #[must_use]
+    pub fn new(gas_usages: &[u128]) -> Self {
```
Comment (Arcticae) [inline crates/forge-runner/src/test_case_summary.rs:19]: Can we perform a breakdown into functions instead of having this all in construction?

**Review by Arcticae — COMMENTED** [review]
Nice job m8

File: crates/forge-runner/src/lib.rs [hunk 8]
```diff
@@ -245,21 +246,19 @@ impl TestCaseSummary<Fuzzing> {
                 test_statistics: (),
             } => {
                 let runs = results.len();
-                let gas_usages_vec = results
+                let gas_usages_vec: Vec<u128> = results
```
Comment (piotmag769) [inline crates/forge-runner/src/lib.rs:249]: I know it's out of scope, but why is this method implemented here and not in `test_case_summary.rs` 🤔

File: crates/forge-runner/src/test_case_summary.rs [hunk 9]
```diff
@@ -3,13 +3,47 @@ use crate::expected_result::{ExpectedPanicValue, ExpectedTestResult};
...
     pub min: u128,
     pub max: u128,
+    pub mean: u128,
+    pub std_deviation: u128,
 }
+
+impl GasStatistics {
+    #[must_use]
+    pub fn new(gas_usages: &[u128]) -> Self {
+        let mean = GasStatistics::mean(gas_usages);
```
Comment (piotmag769) [inline crates/forge-runner/src/test_case_summary.rs:20]: [nit] `Self::mean`

File: crates/forge-runner/src/test_case_summary.rs [hunk 10]
```diff
@@ -3,13 +3,47 @@ use crate::expected_result::{ExpectedPanicValue, ExpectedTestResult};
...
+
+        GasStatistics {
+            min: gas_usages.iter().min().copied().unwrap(),
+            max: gas_usages.iter().max().copied().unwrap(),
+            mean,
+            std_deviation: GasStatistics::std_deviation(mean, gas_usages),
```
Comment (piotmag769) [inline crates/forge-runner/src/test_case_summary.rs:26]: [nit] `Self::std_deviation`

File: crates/forge-runner/src/test_case_summary.rs [hunk 11]
```diff
@@ -3,13 +3,47 @@ use crate::expected_result::{ExpectedPanicValue, ExpectedTestResult};
...
+
+    fn mean(gas_usages: &[u128]) -> u128 {
+        gas_usages.iter().sum::<u128>() / gas_usages.len() as u128
+    }
+
+    fn std_deviation(mean: u128, gas_usages: &[u128]) -> u128 {
+        let sum_squared_diff = gas_usages
+            .iter()
+            .map(|&x| {
+                let abs_difference = if x > mean { x - mean } else { mean - x };
```
Comment (piotmag769) [inline crates/forge-runner/src/test_case_summary.rs:38]: https://doc.rust-lang.org/std/primitive.u128.html#method.abs_diff

**Review by piotmag769 — APPROVED** [review]

File: crates/forge-runner/src/test_case_summary.rs [hunk 12]
```diff
@@ -121,6 +152,58 @@ impl TestCaseSummary<Fuzzing> {
             _ => None,
         }
     }
+
+    #[must_use]
+    pub fn from(results: Vec<TestCaseSummary<Single>>) -> TestCaseSummary<Fuzzing> {
```
Comment (piotmag769) [inline crates/forge-runner/src/test_case_summary.rs:157]: [nit] -> `Self`

File: crates/forge-runner/src/test_case_summary.rs [hunk 13]
```diff
@@ -3,13 +3,44 @@ use crate::expected_result::{ExpectedPanicValue, ExpectedTestResult};
...
+use num_integer::Roots;
 use std::option::Option;
 
-#[derive(Debug, PartialEq, Clone)]
+#[derive(Debug, PartialEq, Clone, Default)]
 pub struct GasStatistics {
     pub min: u128,
     pub max: u128,
+    pub mean: u128,
+    pub std_deviation: u128,
```
Comment (Arcticae) [inline crates/forge-runner/src/test_case_summary.rs:14]: Why the fields if this value is calculated? Could we refactor that and use gas_usages only instead?

**Review by Arcticae — APPROVED** [review]
LGTM

File: crates/forge-runner/src/test_case_summary.rs (hunk 13 above)
Comment (piotmag769) [reply crates/forge-runner/src/test_case_summary.rs:14]: I think it is better to have the results here, we don't need `gas_usages` later on

File: docs/src/testing/gas-etimation.md [hunk 14]
```diff
@@ -7,9 +7,18 @@ and gas-efficient.
...
 ```shell
-[PASS] tests::simple_test, gas: ~1
+[PASS] tests::simple_test, gas = ~1
+```
+
+### Fuzzed Tests
+
+While using the fuzzing feature additional gas statistics will be displayed:
+```shell
+[PASS] tests::fuzzing_test (fuzzer runs = 256), gas = (max = ~126, min = ~1, mean = ~65, std deviation = ~37)
```
Comment (MaksymilianDemitraszek) [inline docs/src/testing/gas-etimation.md:21]: Why fuzzer runs is in parenthesis and gas is not?

Comment (MaksymilianDemitraszek) [reply docs/src/testing/gas-etimation.md:21]: Why mean and std_deviation are integers?

File: crates/forge-runner/src/test_case_summary.rs (hunk 13 above)
Comment (MaksymilianDemitraszek) [reply crates/forge-runner/src/test_case_summary.rs:14]: I think it is better to explicitly define what information runner returns, thus keep those fields here. I don't see why this is a problem 🤔 . It is not a concern of the cli to calculate the statistics imo.

Comment (MaksymilianDemitraszek) [inline crates/forge-runner/src/test_case_summary.rs:14]: Why those are integers?

File: crates/forge-runner/src/test_case_summary.rs [hunk 15]
```diff
@@ -3,13 +3,44 @@ use crate::expected_result::{ExpectedPanicValue, ExpectedTestResult};
...
+    }
+
+    fn mean(gas_usages: &[u128]) -> u128 {
+        gas_usages.iter().sum::<u128>() / gas_usages.len() as u128
+    }
+
+    fn std_deviation(mean: u128, gas_usages: &[u128]) -> u128 {
+        let sum_squared_diff = gas_usages
+            .iter()
+            .map(|&x| x.abs_diff(mean).pow(2))
```
Comment (cptartur) [inline crates/forge-runner/src/test_case_summary.rs:37]: Should it be an absolute difference and not just a difference?

File: crates/forge-runner/src/test_case_summary.rs [hunk 16]
```diff
@@ -121,6 +152,58 @@ impl TestCaseSummary<Fuzzing> {
...
+                name,
+                msg,
+                arguments,
+                gas_info: _,
+                test_statistics: (),
+            } => {
+                let runs = results.len();
+                let gas_usages_vec: Vec<u128> = results
+                    .into_iter()
+                    .filter(|item| matches!(item, TestCaseSummary::Passed { .. }))
```
Comment (cptartur) [inline crates/forge-runner/src/test_case_summary.rs:175]: Do we need this line? From my understanding, they should all be `Passed` if `last` is passed.

**Pull request merged**